required/latex-lab/latex-lab-amsmath.dtx

```diff
@@ -75,87 +76,697 @@
+%    \end{macrocode}
+% \end{macro}
+% \end{macro}
+%
+% \subsubsection{align \& friends}
+% 
+% Most display environment uses a common command for the end which contains
+% the luamml socket to finalize the mtable. 
+% \begin{macro}{\common@align@ending}
+%    \begin{macrocode}
+\def\common@align@ending {
+  \math@cr \black@\totwidth@
+  \UseExpandableTaggingSocket {math/luamml/mtable/finalize} {align}
+  \egroup
+  \ifingather@
+    \restorealignstate@
+    \egroup
+    \nonumber
+    \ifnum0=`{\fi\iffalse}\fi
+  \else
+    $$%
+  \fi
+  \ignorespacesafterend
+}
+%    \end{macrocode}
+% \end{macro}
+% Now we redefine the display alignments to use these ending
+% 
+%    \begin{macrocode}
+\renewenvironment{align}{%
+  \start@align\@ne\st@rredfalse\m@ne
+}{%
+  \common@align@ending
+}
+\renewenvironment{align*}{%
+  \start@align\@ne\st@rredtrue\m@ne
+}{%
+  \common@align@ending
+}
+\renewenvironment{alignat}{%
+  \start@align\z@\st@rredfalse
+}{%
+  \common@align@ending
+}
+\renewenvironment{alignat*}{%
+  \start@align\z@\st@rredtrue
+}{%
+  \common@align@ending
+}
+\renewenvironment{flalign}{%
+  \start@align\tw@\st@rredfalse\m@ne
+}{%
+  \common@align@ending
+}
+\renewenvironment{flalign*}{%
+  \start@align\tw@\st@rredtrue\m@ne
+}{%
+  \common@align@ending
+}
+\renewenvironment{xalignat}{%
+  \start@align\@ne\st@rredfalse
+}{%
+  \common@align@ending
+}
+\renewenvironment{xalignat*}{%
+  \start@align\@ne\st@rredtrue
+}{%
+  \common@align@ending
+}
+\renewenvironment{xxalignat}{%
+  \start@align\tw@\st@rredtrue
+}{%
+  \common@align@ending
+}
+%    \end{macrocode}
+% And register these environments for the math grabbing
+% 
+%    \begin{macrocode}
+  \math_register_halign_env:nn {align}{}
+  \math_register_halign_env:nn {align*}{}
+  \math_register_halign_env:nn {alignat}{}
+  \math_register_halign_env:nn {alignat*}{}
+  \math_register_halign_env:nn {flalign}{}
+  \math_register_halign_env:nn {flalign*}{}
+  \math_register_halign_env:nn {xalignat}{}
+  \math_register_halign_env:nn {xalignat*}{}
+  \math_register_halign_env:nn {xxalignat}{}
+%    \end{macrocode}
+%
+% The align preamble (used in \cs{align@}) needs code for luamml 
+% to save the cells.
+% \begin{macro}{\align@preamble}
+%    \begin{macrocode}
+\def\align@preamble{%
+   &\hfil
+    \strut@
+    \setboxz@h
+      {
+        \@lign
+         $
+          \m@th\displaystyle{##}         
+         \ifmeasuring@
+          \luamml_ignore:
+         \else
+          \UseTaggingSocket{math/luamml/save/nNn}{ {} \displaystyle {mtd}}
+         \fi 
+         $
+      }%
+    \ifmeasuring@
+      \savefieldlength@
+    \else
+      \UseTaggingSocket{math/luamml/mtable/finalizecol}{box}
+    \fi
+    \set@field
+    \tabskip\z@skip
+   &\setboxz@h
+      {
+        \@lign
+         $
+           \m@th\displaystyle{{}##}        
+           \ifmeasuring@
+           \luamml_ignore:
+           \else
+           \UseTaggingSocket{math/luamml/save/nNn}{ {} \displaystyle {mtd}}
```

Comment:
Any chance for a more descriptive name for this socket? (instead of save/nNn)
Better:  {mtd} }

Reply:
The socket is a simple wrapper around `\luamml_save:nNn` and has no additional intrinsic meaning.  (I unified the spaces.)